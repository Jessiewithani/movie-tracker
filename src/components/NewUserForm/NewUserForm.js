import React, { Component } from 'react';
import { Redirect } from 'react-router-dom'
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import './NewUserForm.scss'
import { saveUser } from '../../actions'
import { postNewUser } from '../../apiCalls/apiCalls';

export class NewUserForm extends Component {
  constructor() {
    super()
    this.state = {
      newName: '',
      newEmail: '',
      newPassword: '',
      status: null,
      createEmailError: ''
    }
  }

  handleChange = (e) => {
    this.setState({[e.target.name]: e.target.value});
  }

  clearNewUserInputs = () => {
    this.setState({newName: '', newEmail: '', newPassword: ''})
  }

  submitNewUserInfo = async e => {
    e.preventDefault();
    const { saveUser } = this.props;
    const { newName, newEmail, newPassword } = this.state;
    const newUser = {name: newName, email: newEmail, password: newPassword};

    try {
      const response = await postNewUser(newUser);
      this.setState({ status: response.status }); 
      if (this.state.status === 201) {
        saveUser(newName, newEmail, newPassword);
      }
    } catch(error) {
      throw Error(error)
    }
    this.clearNewUserInputs()
  }
  
  render() {
    if (this.state.status === 201) {
      return <Redirect to='/login' />
    } else if(this.state.status === 500) {
      this.setState({ createEmailError: "* this email already exists *"  })
    }
 
    return(
      <div className="new-user-container">
        <section className="create-account">
          <form>
            <input className="new-user-input" type="text" placeholder="insert your name" name="newName" value={this.state.newName || ''} onChange={this.handleChange} />
            <input className="email-error new-user-input" type="email" placeholder="insert your e-mail" name="newEmail" value={this.state.newEmail || ''} onChange={this.handleChange} />
            <h4 style={{color: "red"}}>{this.state.createEmailError}</h4>
            <input className="new-user-input" type="password" placeholder="insert your password" name="newPassword" value={this.state.newPassword || ''} onChange={this.handleChange} />
            <button className="create-user-button a-create-button" onClick={(e) => this.submitNewUserInfo(e)}>SUBMIT</button>
          </form>
        </section>
      </div>
    )
  }
}

const mapStateToProps = ({ user }) => ({
  user
})

const mapDispatchToProps = dispatch => (
  bindActionCreators({
    saveUser
  }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(NewUserForm);