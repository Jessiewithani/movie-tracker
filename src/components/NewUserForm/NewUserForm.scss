.create-account {
    background: url('https://designshack.net/wp-content/uploads/best-subtle-black-white-background-textures.jpg');
    justify-content: center;
    align-items: center;
    display: flex;

    height: 100vh;

}

form {
    // background-color: black;
    height: 500px;
    width: 700px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    justify-content: space-evenly;
}

input {
    font-size: 2.5em;
    color: white;
    padding-left: 15px;
    border: none;
    border-bottom: solid black 5px;
    background-color: rgba(0, 0, 0, 0);
}

.create-user-button {
    font-size: 2em;
    color: lightgrey;
    font-weight: bold;
    height: 2.5em;
    width: 700px;
    background-color: rgba(0,0,0,0.5)
}

a.a-create-button:hover {
    background-color: white;
}







