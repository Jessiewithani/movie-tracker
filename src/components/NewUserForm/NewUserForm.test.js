import NewUserForm from './NewUserForm';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { Redirect } from 'react-router-dom'
import { MemoryRouter } from 'react-router';



describe('NewUserForm', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <NewUserForm
      />
    )
  })

  it('should match the snapshot with all data passed in', () => {
    expect(wrapper).toMatchSnapshot()
  });
  it('should call submitNewUserInfo when the button is clicked', () => {

    wrapper.instance().submitNewUserInfo = jest.fn()
    const mockEvent = { preventDefault: jest.fn() }

    wrapper.instance().forceUpdate();

    wrapper.find('button').simulate('click', mockEvent)

    expect(wrapper.instance().submitNewUserInfo).toHaveBeenCalled()
  });
  it('should update state when handleChange is invoked', () => {
    const mockEvent = {target: {name: 'name', value: 'Eric'}}
    const expected = 'Eric'

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('name')).toEqual(expected)
  });
  it('should redirect to login if the user is successfully added', () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={['/new-user']}>
        <Redirect to={'/login'}/>
      </MemoryRouter>
    )

    expect(wrapper.find(Redirect)).toHaveLength(1)
  })
})