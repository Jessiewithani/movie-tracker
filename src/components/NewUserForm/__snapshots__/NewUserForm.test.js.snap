// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewUserForm should match the snapshot with all data passed in 1`] = `
<div
  className="new-user-container"
>
  <section
    className="create-account"
  >
    <form>
      <input
        className="new-user-input"
        name="newName"
        onChange={[Function]}
        placeholder="insert your name"
        type="text"
        value=""
      />
      <input
        className="new-user-input"
        name="newEmail"
        onChange={[Function]}
        placeholder="insert your e-mail"
        type="email"
        value=""
      />
      <input
        className="new-user-input"
        name="newPassword"
        onChange={[Function]}
        placeholder="insert your password"
        type="password"
        value=""
      />
      <a
        className="a-create-button"
      >
        <button
          className="create-user-button"
          onClick={[Function]}
        >
          SUBMIT
        </button>
      </a>
    </form>
  </section>
</div>
`;
